<a href="https://colab.research.google.com/github/muhxz3/Messy-Employee-Dataset-Analysis/blob/main/Cleaned_Employee_Dataset_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Identifying Missing Values

In [437]:
import pandas as pd

In [438]:
# Loading the dataset
data = pd.read_csv("/content/Day11_Messy_Company_Employee_Dataset.csv")
data

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
0,EMP0098,Ananya Nair,NaN,Data Scientist,48.0,Other,108371.0,13.2,2017-06-13,Pune,5.0,Office
1,EMP0046,Faizan Khan,Marketing,Marketing Manager,31.0,Female,108824.0,12.1,2021-09-15,Mumbai,2.0,Office
2,EMP0017,Saira Malik,Human Resources,HR Manager,28.0,Female,119400.0,13.2,2021-02-05,Bengaluru,3.0,Remote
3,EMP0105,Neha Menon,Engineering,Software Engineer,30.0,Female,111847.0,NaN,2018-08-05,Jaipur,4.0,Hybrid
4,EMP0143,Hiba Menon,Sales,Sales Manager,25.0,Female,78677.0,10.9,2018-07-24,Hyderabad,4.0,Hybrid
...,...,...,...,...,...,...,...,...,...,...,...,...
152,EMP0132,Manya Rao,Marketing,Content Strategist,30.0,Female,49522.0,0.6,2021-11-25,Delhi,2.0,Remote
153,EMP0073,Karan Kapoor,Customer Success,Customer Success Manager,42.0,MALE,43331.0,1.1,2018-03-23,Jaipur,3.0,Hybrid
154,EMP0124,Saira Kapoor,Engineering,QA Engineer,33.0,Male,94812.0,7.2,2024-06-12,Jaipur,4.0,Hybrid
155,EMP0020,Sara Ali,Sales,Sales Executive,22.0,Female,51745.0,2.4,2020-11-19,NaN,2.0,Office


In [439]:
# identifying missing values
data.isna()

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
0,False,False,True,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,True,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...
152,False,False,False,False,False,False,False,False,False,False,False,False
153,False,False,False,False,False,False,False,False,False,False,False,False
154,False,False,False,False,False,False,False,False,False,False,False,False
155,False,False,False,False,False,False,False,False,False,True,False,False


In [440]:
data.isna().sum()

,0
Employee_ID,0
Employee_Name,0
Department,5
Job_Title,0
Age,4
Gender,5
Annual_Salary,5
Experience_Years,3
Joining_Date,0
City,5


In [441]:
data.isna().sum().sum()

np.int64(32)

- There are 32 Null/Missing Values

In [442]:
data.dtypes

,0
Employee_ID,object
Employee_Name,object
Department,object
Job_Title,object
Age,float64
Gender,object
Annual_Salary,float64
Experience_Years,float64
Joining_Date,object
City,object


In [443]:
data.shape

(157, 12)

In [444]:
# dropping rows with null values in Gender
data = data.dropna(subset=['Gender'])
data

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
0,EMP0098,Ananya Nair,NaN,Data Scientist,48.0,Other,108371.0,13.2,2017-06-13,Pune,5.0,Office
1,EMP0046,Faizan Khan,Marketing,Marketing Manager,31.0,Female,108824.0,12.1,2021-09-15,Mumbai,2.0,Office
2,EMP0017,Saira Malik,Human Resources,HR Manager,28.0,Female,119400.0,13.2,2021-02-05,Bengaluru,3.0,Remote
3,EMP0105,Neha Menon,Engineering,Software Engineer,30.0,Female,111847.0,NaN,2018-08-05,Jaipur,4.0,Hybrid
4,EMP0143,Hiba Menon,Sales,Sales Manager,25.0,Female,78677.0,10.9,2018-07-24,Hyderabad,4.0,Hybrid
...,...,...,...,...,...,...,...,...,...,...,...,...
152,EMP0132,Manya Rao,Marketing,Content Strategist,30.0,Female,49522.0,0.6,2021-11-25,Delhi,2.0,Remote
153,EMP0073,Karan Kapoor,Customer Success,Customer Success Manager,42.0,MALE,43331.0,1.1,2018-03-23,Jaipur,3.0,Hybrid
154,EMP0124,Saira Kapoor,Engineering,QA Engineer,33.0,Male,94812.0,7.2,2024-06-12,Jaipur,4.0,Hybrid
155,EMP0020,Sara Ali,Sales,Sales Executive,22.0,Female,51745.0,2.4,2020-11-19,NaN,2.0,Office


In [445]:
data.shape

(152, 12)

In [446]:
data.isna().sum()

,0
Employee_ID,0
Employee_Name,0
Department,5
Job_Title,0
Age,4
Gender,0
Annual_Salary,5
Experience_Years,3
Joining_Date,0
City,5


In [447]:
data

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
0,EMP0098,Ananya Nair,NaN,Data Scientist,48.0,Other,108371.0,13.2,2017-06-13,Pune,5.0,Office
1,EMP0046,Faizan Khan,Marketing,Marketing Manager,31.0,Female,108824.0,12.1,2021-09-15,Mumbai,2.0,Office
2,EMP0017,Saira Malik,Human Resources,HR Manager,28.0,Female,119400.0,13.2,2021-02-05,Bengaluru,3.0,Remote
3,EMP0105,Neha Menon,Engineering,Software Engineer,30.0,Female,111847.0,NaN,2018-08-05,Jaipur,4.0,Hybrid
4,EMP0143,Hiba Menon,Sales,Sales Manager,25.0,Female,78677.0,10.9,2018-07-24,Hyderabad,4.0,Hybrid
...,...,...,...,...,...,...,...,...,...,...,...,...
152,EMP0132,Manya Rao,Marketing,Content Strategist,30.0,Female,49522.0,0.6,2021-11-25,Delhi,2.0,Remote
153,EMP0073,Karan Kapoor,Customer Success,Customer Success Manager,42.0,MALE,43331.0,1.1,2018-03-23,Jaipur,3.0,Hybrid
154,EMP0124,Saira Kapoor,Engineering,QA Engineer,33.0,Male,94812.0,7.2,2024-06-12,Jaipur,4.0,Hybrid
155,EMP0020,Sara Ali,Sales,Sales Executive,22.0,Female,51745.0,2.4,2020-11-19,NaN,2.0,Office


In [448]:
# Filling Annual Salary
data['Annual_Salary'] = data['Annual_Salary'].fillna(data['Annual_Salary'].mean())
data

/tmp/ipykernel_1993/1895039504.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['Annual_Salary'] = data['Annual_Salary'].fillna(data['Annual_Salary'].mean())


,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
0,EMP0098,Ananya Nair,NaN,Data Scientist,48.0,Other,108371.0,13.2,2017-06-13,Pune,5.0,Office
1,EMP0046,Faizan Khan,Marketing,Marketing Manager,31.0,Female,108824.0,12.1,2021-09-15,Mumbai,2.0,Office
2,EMP0017,Saira Malik,Human Resources,HR Manager,28.0,Female,119400.0,13.2,2021-02-05,Bengaluru,3.0,Remote
3,EMP0105,Neha Menon,Engineering,Software Engineer,30.0,Female,111847.0,NaN,2018-08-05,Jaipur,4.0,Hybrid
4,EMP0143,Hiba Menon,Sales,Sales Manager,25.0,Female,78677.0,10.9,2018-07-24,Hyderabad,4.0,Hybrid
...,...,...,...,...,...,...,...,...,...,...,...,...
152,EMP0132,Manya Rao,Marketing,Content Strategist,30.0,Female,49522.0,0.6,2021-11-25,Delhi,2.0,Remote
153,EMP0073,Karan Kapoor,Customer Success,Customer Success Manager,42.0,MALE,43331.0,1.1,2018-03-23,Jaipur,3.0,Hybrid
154,EMP0124,Saira Kapoor,Engineering,QA Engineer,33.0,Male,94812.0,7.2,2024-06-12,Jaipur,4.0,Hybrid
155,EMP0020,Sara Ali,Sales,Sales Executive,22.0,Female,51745.0,2.4,2020-11-19,NaN,2.0,Office


In [449]:
data.isna().sum()

,0
Employee_ID,0
Employee_Name,0
Department,5
Job_Title,0
Age,4
Gender,0
Annual_Salary,0
Experience_Years,3
Joining_Date,0
City,5


In [450]:
# filling missing experience years with zero
data["Experience_Years"] = data["Experience_Years"].fillna(0)


/tmp/ipykernel_1993/1310600910.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data["Experience_Years"] = data["Experience_Years"].fillna(0)


In [451]:
data.isna().sum()

,0
Employee_ID,0
Employee_Name,0
Department,5
Job_Title,0
Age,4
Gender,0
Annual_Salary,0
Experience_Years,0
Joining_Date,0
City,5


In [452]:
data.dtypes

,0
Employee_ID,object
Employee_Name,object
Department,object
Job_Title,object
Age,float64
Gender,object
Annual_Salary,float64
Experience_Years,float64
Joining_Date,object
City,object


In [453]:
data["Age"]

,Age
0,48.0
1,31.0
2,28.0
3,30.0
4,25.0
...,...
152,30.0
153,42.0
154,33.0
155,22.0


In [454]:
# dropping rows missing Age Values
data = data.dropna(subset=["Age"])

In [455]:
data.shape

(148, 12)

In [456]:
data["Performance_Score"]

,Performance_Score
0,5.0
1,2.0
2,3.0
3,4.0
4,4.0
...,...
152,2.0
153,3.0
154,4.0
155,2.0


In [457]:
# filling missing performance score with its mean
data["Performance_Score"] = data["Performance_Score"].fillna(data["Performance_Score"].mean())

/tmp/ipykernel_1993/3523187665.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data["Performance_Score"] = data["Performance_Score"].fillna(data["Performance_Score"].mean())


In [458]:
data.isna().sum()

,0
Employee_ID,0
Employee_Name,0
Department,5
Job_Title,0
Age,0
Gender,0
Annual_Salary,0
Experience_Years,0
Joining_Date,0
City,5


In [459]:
data["Work_Mode"]

,Work_Mode
0,Office
1,Office
2,Remote
3,Hybrid
4,Hybrid
...,...
152,Remote
153,Hybrid
154,Hybrid
155,Office


In [460]:
data["Work_Mode"].values

array(['Office', 'Office', 'Remote', 'Hybrid', 'Hybrid', 'Hybrid',
       'Office', 'Hybrid', 'Remote', 'Remote', nan, 'Hybrid', 'Remote',
       'Hybrid', 'Remote', 'Remote', 'Remote', 'Hybrid', 'Office',
       'Hybrid', 'REMOTE', 'Hybrid', 'Remote', 'Office', 'Office',
       'Office', 'Remote', 'Office', 'Hybrid', 'Office', 'Remote',
       'Office', 'Hybrid', 'Office', 'Remote', 'Remote', 'Hybrid',
       'Office', 'Office', 'Office', 'Hybrid', 'Office', 'Remote',
       'Remote', 'Hybrid', 'Hybrid', 'Remote', 'Hybrid', 'Remote',
       'Remote', 'Office', 'Office', 'Office', 'Remote', 'Office', nan,
       'Hybrid', 'Hybrid', 'Hybrid', 'Hybrid', 'Remote', 'Hybrid',
       'Hybrid', 'Hybrid', 'Remote', 'Remote', 'Remote', 'Hybrid',
       'Hybrid', 'Hybrid', 'Office', 'Remote', 'Hybrid', 'Office',
       'Office', 'Office', 'Office', 'Hybrid', 'Office', 'Hybrid',
       'Remote', 'Remote', 'Office', 'Remote', 'Hybrid', 'Office',
       'Hybrid', 'Office', 'Remote', 'Hybrid', 'remo

In [461]:
# dropping rows with missing work modes
data = data.dropna(subset=["Work_Mode"])

In [462]:
data.shape

(146, 12)

In [463]:
# filling missinf city to unknown
data["City"] = data["City"].fillna("Unknown")

In [464]:
# dropping missing department values
data = data.dropna(subset=["Department"])

In [465]:
data.isna().sum()

,0
Employee_ID,0
Employee_Name,0
Department,0
Job_Title,0
Age,0
Gender,0
Annual_Salary,0
Experience_Years,0
Joining_Date,0
City,0


- all the missing values have been corrected


In [466]:
data

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
1,EMP0046,Faizan Khan,Marketing,Marketing Manager,31.0,Female,108824.0,12.1,2021-09-15,Mumbai,2.0,Office
2,EMP0017,Saira Malik,Human Resources,HR Manager,28.0,Female,119400.0,13.2,2021-02-05,Bengaluru,3.0,Remote
3,EMP0105,Neha Menon,Engineering,Software Engineer,30.0,Female,111847.0,0.0,2018-08-05,Jaipur,4.0,Hybrid
4,EMP0143,Hiba Menon,Sales,Sales Manager,25.0,Female,78677.0,10.9,2018-07-24,Hyderabad,4.0,Hybrid
5,EMP0117,Zoya Kapoor,Finance,Finance Manager,42.0,Male,128567.0,15.4,2017-08-02,Chennai,4.0,Hybrid
...,...,...,...,...,...,...,...,...,...,...,...,...
152,EMP0132,Manya Rao,Marketing,Content Strategist,30.0,Female,49522.0,0.6,2021-11-25,Delhi,2.0,Remote
153,EMP0073,Karan Kapoor,Customer Success,Customer Success Manager,42.0,MALE,43331.0,1.1,2018-03-23,Jaipur,3.0,Hybrid
154,EMP0124,Saira Kapoor,Engineering,QA Engineer,33.0,Male,94812.0,7.2,2024-06-12,Jaipur,4.0,Hybrid
155,EMP0020,Sara Ali,Sales,Sales Executive,22.0,Female,51745.0,2.4,2020-11-19,Unknown,2.0,Office


# Identifying Inconsistent values

In [467]:
data.groupby("Department")["Department"].value_counts()

,count
Department,
Customer Success,20
Data & Analytics,14
ENGINEERING,1
Engineering,19
Finance,18
Human Resources,15
Marketing,20
Operations,15
Sales,18


In [468]:
data.loc[data["Department"].isin(["ENGINEERING", "engineering"]), "Department"] = "Engineering"

In [469]:
data.groupby("Department")["Department"].value_counts()

,count
Department,
Customer Success,20
Data & Analytics,14
Engineering,21
Finance,18
Human Resources,15
Marketing,20
Operations,15
Sales,18


In [470]:
data.groupby("Job_Title")["Job_Title"].value_counts()

,count
Job_Title,
Accountant,5
BI Analyst,3
Business Development Executive,4
Content Strategist,7
Customer Success Executive,5
Customer Success Manager,7
Data Scientist,11
Finance Manager,4
Financial Analyst,9


In [471]:
data.groupby("Gender")["Gender"].value_counts()

,count
Gender,
Female,63
MALE,1
Male,72
Other,4
female,1


In [472]:
data.loc[data["Gender"]=='female',"Gender"] = 'Female'
data.loc[data["Gender"]=='MALE',"Gender"] = 'Male'

In [473]:
data.groupby("Gender")["Gender"].value_counts()

,count
Gender,
Female,64
Male,73
Other,4


In [474]:
data.groupby("City")["City"].value_counts()

,count
City,
Bengaluru,12
Chandigarh,21
Chennai,9
DELHI,1
Delhi,10
Delhi,1
Hyderabad,12
Jaipur,17
Kolkata,8


In [475]:
data.loc[data["City"].isin(["DELHI","delhi","Delhi "]),"City"] = "Delhi"

In [476]:
data.groupby("City")["City"].value_counts()

,count
City,
Bengaluru,12
Chandigarh,21
Chennai,9
Delhi,13
Hyderabad,12
Jaipur,17
Kolkata,8
Mumbai,14
Pune,20


In [477]:
data.groupby("Work_Mode")["Work_Mode"].value_counts()

,count
Work_Mode,
Hybrid,46
Office,41
REMOTE,1
Remote,52
remote,1


In [478]:
data.loc[data["Work_Mode"].isin(["REMOTE","remote"]),"Work_Mode"] = 'Remote'

In [479]:
data.groupby("Work_Mode")["Work_Mode"].value_counts()

,count
Work_Mode,
Hybrid,46
Office,41
Remote,54


In [480]:
data.duplicated().sum()

np.int64(7)

In [481]:
data.shape

(141, 12)

In [482]:
data = data.drop_duplicates()
data

,Employee_ID,Employee_Name,Department,Job_Title,Age,Gender,Annual_Salary,Experience_Years,Joining_Date,City,Performance_Score,Work_Mode
1,EMP0046,Faizan Khan,Marketing,Marketing Manager,31.0,Female,108824.0,12.1,2021-09-15,Mumbai,2.0,Office
2,EMP0017,Saira Malik,Human Resources,HR Manager,28.0,Female,119400.0,13.2,2021-02-05,Bengaluru,3.0,Remote
3,EMP0105,Neha Menon,Engineering,Software Engineer,30.0,Female,111847.0,0.0,2018-08-05,Jaipur,4.0,Hybrid
4,EMP0143,Hiba Menon,Sales,Sales Manager,25.0,Female,78677.0,10.9,2018-07-24,Hyderabad,4.0,Hybrid
5,EMP0117,Zoya Kapoor,Finance,Finance Manager,42.0,Male,128567.0,15.4,2017-08-02,Chennai,4.0,Hybrid
...,...,...,...,...,...,...,...,...,...,...,...,...
151,EMP0050,Alina Patel,Finance,Financial Analyst,22.0,Male,74903.0,7.6,2020-01-18,Pune,2.0,Remote
152,EMP0132,Manya Rao,Marketing,Content Strategist,30.0,Female,49522.0,0.6,2021-11-25,Delhi,2.0,Remote
153,EMP0073,Karan Kapoor,Customer Success,Customer Success Manager,42.0,Male,43331.0,1.1,2018-03-23,Jaipur,3.0,Hybrid
154,EMP0124,Saira Kapoor,Engineering,QA Engineer,33.0,Male,94812.0,7.2,2024-06-12,Jaipur,4.0,Hybrid


In [483]:
data.duplicated().sum()

np.int64(0)

# Verification
- There were 157 rows in the beginning. After cleaning and droppping duplicates there are 134 rows.
- Some of the missing values have been filled and some rows are dropped.

In [485]:
data.to_csv("Cleaned_Employee_Dataset.csv")